src/armature/Animation.ts

```diff
@@ -0,0 +1,178 @@
+import { Node } from './Node';
+import { Transformation } from './Transformation';
+
+import { remove } from 'lodash';
+
+type AnimationDescription = {
+    node: Node;
+    to: (node: Node) => void;
+    finalTransform: Transformation | null;
+    curve: string;
+    start: number;
+    duration: number;
+    times: number;
+    repeatDelay: number;
+    lastTimeInCycle: number;
+    lastCycle: number;
+};
+
+type AnimationDescriptionParams = {
+    node: Node;
+    to: (node: Node) => void;
+    curve?: string;
+    start?: number;
+    duration: number;
+    times?: number;
+    repeatDelay?: number;
+};
+
+export namespace Animation {
+    const queue: AnimationDescription[] = [];
+    const current: AnimationDescription[] = [];
+
+    /**
+     * Clears all current animations.
+     */
+    export function resetAll() {
+        queue.length = 0;
+        current.length = 0;
+    }
+
+    /**
+     * @returns {number} The current time, in milliseconds since the epoch.
+     */
+    export function now(): number {
+        return new Date().getTime();
+    }
+
+    /**
+     * Creates and queues up an animation with the given parameters.
+     *
+     * @param {AnimationDescriptionParams} params The information about the new animation.
+     */
+    export function create(params: AnimationDescriptionParams) {
+        const animation = {
+            node: params.node,
+            to: params.to,
+            finalTransform: null,
+            curve: params.curve !== undefined ? params.curve : 'linear',
+            start: params.start !== undefined ? params.start : Animation.now(),
+            duration: params.duration,
+            times: params.times !== undefined ? params.times : 1,
+            repeatDelay: params.repeatDelay !== undefined ? params.repeatDelay : 0,
+            lastTimeInCycle: 0,
+            lastCycle: -1
+        };
+
+        enqueue(animation);
+    }
+
+    /**
+     * Updates the state of all animations according to the current time.
+     */
+    export function tick() {
+        const currentTime = Animation.now();
+
+        dequeueNewAnimations(currentTime);
+        applyCurrentAnimations(currentTime);
+        removeFinishedAnimations(currentTime);
+    }
+
+    /**
+     * Adds an animation to the queue, ensuring that the queue is kept in order of start time.
+     */
+    function enqueue(animation: AnimationDescription) {
+        // TODO: binary search instead of insert + sort
+        queue.push(animation);
+        queue.sort((a: AnimationDescription, b: AnimationDescription) => {
+            if (a.start < b.start) {
+                return -1;
+            }
+            if (a.start > b.start) {
```

Comment:
Is there a reason this isn't an `else if`, and part of the previous `if` statement?

```typescript
if (a.start < b.start) {
  return -1;
} else if (a.start > b.start) {
  return 1;
} else {
  return 0;
}
```

Reply:
We return in the middle so it's just a question of readability I guess since the code is the same. I can change that